syntax = "proto2";

package raft;

option cc_generic_services = true;

message LogEntry {
    required int64 term = 1;
    required bytes command = 2;
}

message RequestVoteRequest {
    required int64 term = 1;
    required int64 candidate_id = 2;
    required int64 last_log_index = 3;
    required int64 last_log_term = 4;
}

message RequestVoteResponse {
    required int64 term = 1;
    required bool vote_granted = 2;
}

message AppendEntriesRequest {
    required int64 term = 1;
    required int64 leader_id = 2;
    required int64 prev_log_index = 3;
    required int64 prev_log_term = 4;
    repeated LogEntry entries = 5;
    required int64 leader_commit = 6;
}

message AppendEntriesResponse {
    required int64 term = 1;
    required bool success = 2;
}

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}